riscv/src/continuations.rs

```diff
@@ -155,45 +156,127 @@ fn sanity_check(main_machine: &Machine, field: KnownField) {
+    let mut initial_memory = if let Some(expr) = extract_var_from_machine(machine, "initial_memory")
+    {
+        let Expression::ArrayLiteral(_, array) = expr else {
+            panic!("initial_memory is not an array literal");
+        };
+
+        array
+            .items
+            .iter()
+            .map(|entry| {
+                let Expression::Tuple(_, tuple) = entry else {
+                    panic!("initial_memory entry is not a tuple");
+                };
+                assert_eq!(tuple.len(), 2);
+
+                let key = expr_to_u32(&tuple[0]).expect("initial_memory entry key is not a u32");
+                let value =
+                    expr_to_u32(&tuple[1]).expect("initial_memory entry value is not a u32");
+
+                (key, value)
+            })
+            .collect()
+    } else {
         log::warn!("No initial_memory variable found in the machine. Assuming zeroed memory.");
-        return MemoryState::default();
+        MemoryState::default()
     };
 
-    let Expression::ArrayLiteral(_, array) = expr else {
-        panic!("initial_memory is not an array literal");
-    };
+    // Fill the first page with random data to be the salt.
+    // TODO: the random value should be the "hash" of the merkle tree leaf itself,
+    // and the preimage of such hash should be unknown. But to implement that it would
+    // require some inteligence on the MemoryState type.
+    let mut rng = rand::rngs::OsRng;
+    for i in 0..(PAGE_SIZE_BYTES / 4) {
+        initial_memory.insert(prover_data_start + i * 4, rng.gen::<u32>());
+    }
 
-    array
-        .items
-        .iter()
-        .map(|entry| {
-            let Expression::Tuple(_, tuple) = entry else {
-                panic!("initial_memory entry is not a tuple");
-            };
-            assert_eq!(tuple.len(), 2);
-            let Expression::Number(
-                _,
-                Number {
-                    value: key,
-                    type_: None,
-                },
-            ) = &tuple[0]
-            else {
-                panic!("initial_memory entry key is not a number");
-            };
-            let Expression::Number(_, Number { value, type_: None }) = &tuple[1] else {
-                panic!("initial_memory entry value is not a number");
-            };
-
-            (key.try_into().unwrap(), value.try_into().unwrap())
-        })
-        .collect()
+    // Actually fill the prover data
+
+    // Setup an iterator for the addresses to be filled. If the iterator runs out before
+    // we are done, it means that the prover data is too large to fit in the reserved space.
+    let mut word_addr_iter =
+        ((prover_data_start + PAGE_SIZE_BYTES) / 4..prover_data_end / 4).map(|i| i * 4);
+
+    // The first word is the total number of words that will follow in the user data.
+    // We save tha address to write it later, when we know it.
+    let total_word_count_addr = word_addr_iter.next().unwrap();
```

Comment:
maybe `next().unwrap()` => `next().expect(out_of_space_msg)`?

Reply:
or, this could be in its own function that returns an Option/Error (so you can use `?`) and here you just panic with a nicer msg?